apps/hash-frontend/src/pages/[shortname]/[page-slug].page/canvas-page/block-creation-dialog.tsx

```diff
@@ -0,0 +1,162 @@
+import { useMutation } from "@apollo/client";
+import { VersionedUrl } from "@blockprotocol/type-system/dist/cjs-slim/index-slim";
+import {
+  getPageQuery,
+  updatePageContents,
+} from "@local/hash-graphql-shared/queries/page.queries";
+import { HashBlockMeta } from "@local/hash-isomorphic-utils/blocks";
+import { TEXT_ENTITY_TYPE_ID } from "@local/hash-isomorphic-utils/entity-store";
+import { OwnedById } from "@local/hash-subgraph";
+import { useApp } from "@tldraw/editor";
+import { DialogProps } from "@tldraw/tldraw";
+import { useCallback, useState } from "react";
+
+import { BlockSuggester } from "../../../../blocks/page/create-suggester/block-suggester";
+import { usePageContext } from "../../../../blocks/page/page-context";
+import {
+  UpdatePageContentsMutation,
+  UpdatePageContentsMutationVariables,
+} from "../../../../graphql/api-types.gen";
+import { useRouteNamespace } from "../../shared/use-route-namespace";
+import { BlockShape } from "./block-shape";
+import { defaultBlockHeight, defaultBlockWidth } from "./shared";
+
+// An interface for selecting a Block Protocol block and creating the entity it needs
+export const BlockCreationDialog = ({ onClose }: DialogProps) => {
+  const [creatingEntity, setCreatingEntity] = useState(false);
+
+  const app = useApp();
+
+  const { routeNamespace } = useRouteNamespace();
+
+  const { accountId } = routeNamespace ?? {};
+
+  const { pageEntityId } = usePageContext();
+
+  const [updatePageContentsFn] = useMutation<
+    UpdatePageContentsMutation,
+    UpdatePageContentsMutationVariables
+  >(updatePageContents);
+
+  const createBlock = useCallback(
+    async (blockMeta: HashBlockMeta) => {
+      const position = app.getShapesInPage(app.pages[0]!.id).length;
+
+      const blockEntityTypeId = blockMeta.schema as VersionedUrl;
+
+      const width = defaultBlockWidth;
+      const height = defaultBlockHeight;
+
+      const x = app.viewportPageCenter.x - width / 2;
+      const y = app.viewportPageCenter.y - height / 2;
+
+      if (!accountId) {
+        throw new Error(
+          "No accountId available – possibly routeNamespace is not yet loaded",
+        );
+      }
+
+      const { data } = await updatePageContentsFn({
+        variables: {
+          entityId: pageEntityId,
+          actions: [
+            {
+              insertBlock: {
+                componentId: blockMeta.componentId,
+                entity: {
+                  // @todo this should be 'blockEntityTypeId' above but external types are not yet fully supported
+                  entityTypeId: TEXT_ENTITY_TYPE_ID,
```

Comment:
This TODO should be able to be done now as https://github.com/hashintel/hash/pull/2517 has been merged

Reply:
Done in https://github.com/hashintel/hash/pull/2503/commits/8fabc1cc73a152f8d73cae9ece015add83d11ac8